File: paleo_workbench/ui/native_map_canvas.py
"""Qt host canvas that composes native scalar rasters with contours and sample points."""
from __future__ import annotations

from typing import Any

from PySide6.QtCore import QPointF, QRectF, Qt
from PySide6.QtGui import QColor, QImage, QMouseEvent, QPainter, QPen, QWheelEvent
from PySide6.QtWidgets import QWidget

from paleo_workbench.ui import tokens

__all__ = ["NativeMapCanvas"]


class NativeMapCanvas(QWidget):
    """Native Qt map canvas with cached scalar images and pointer pan/zoom."""

    def __init__(self, scene: Any | None = None, parent=None):
        super().__init__(parent)
        self.setObjectName("NativeMapCanvas")
        self.setMinimumSize(240, 180)
        self.setMouseTracking(True)
        self._scene = None
        self._view_extent = (0.0, 0.0, 1.0, 1.0)
        self._image_cache: dict[str, tuple[tuple[int, int], QImage]] = {}
        self._last_drag_pos: QPointF | None = None
        self._scene_change_listener = self.update
        if scene is not None:
            self.set_scene(scene)

    @property
    def scene(self):
        return self._scene

    @property
    def view_extent(self) -> tuple[float, float, float, float]:
        return self._view_extent

    def set_scene(self, scene) -> None:
        if self._scene is not None:
            self._scene.remove_change_listener(self._scene_change_listener)
        self._scene = scene
        self._scene.add_change_listener(self._scene_change_listener)
        self._image_cache.clear()
        self.fit_to_scene()
        self.update()

    def fit_to_scene(self) -> None:
        if self._scene is None:
            self._view_extent = (0.0, 0.0, 1.0, 1.0)
            return
        xmin, ymin, xmax, ymax = self._scene.extent()
        dx = xmax - xmin
        dy = ymax - ymin
        if dx <= 0.0 or dy <= 0.0:
            self._view_extent = (0.0, 0.0, 1.0, 1.0)
            return
        margin = 0.04
        self._view_extent = (
            xmin - dx * margin,
            ymin - dy * margin,
            xmax + dx * margin,
            ymax + dy * margin,
        )

    def set_view_extent(self, extent: tuple[float, float, float, float]) -> None:
        xmin, ymin, xmax, ymax = (float(value) for value in extent)
        if not (xmax > xmin and ymax > ymin):
            raise ValueError("view extent must have positive width and height")
        self._view_extent = (xmin, ymin, xmax, ymax)
        self.update()

    def zoom_to_extent(self, extent: tuple[float, float, float, float]) -> None:
        self.set_view_extent(extent)

    def zoom_by(self, factor: float, center: tuple[float, float] | None = None) -> None:
        if factor <= 0.0:
            raise ValueError("zoom factor must be positive")
        xmin, ymin, xmax, ymax = self._view_extent
        cx, cy = center or ((xmin + xmax) / 2.0, (ymin + ymax) / 2.0)
        self.set_view_extent(
            (
                cx + (xmin - cx) * factor,
                cy + (ymin - cy) * factor,
                cx + (xmax - cx) * factor,
                cy + (ymax - cy) * factor,
            )
        )

    def pan_by_pixels(self, dx: float, dy: float) -> None:
        xmin, ymin, xmax, ymax = self._view_extent
        width = max(1, self.width())
        height = max(1, self.height())
        world_dx = -float(dx) * (xmax - xmin) / width
        world_dy = float(dy) * (ymax - ymin) / height
        self.set_view_extent((xmin + world_dx, ymin + world_dy, xmax + world_dx, ymax + world_dy))

    def _world_to_screen(self, x: float, y: float) -> QPointF:
        xmin, ymin, xmax, ymax = self._view_extent
        return QPointF(
            (x - xmin) * self.width() / (xmax - xmin),
            self.height() - (y - ymin) * self.height() / (ymax - ymin),
        )

    def _screen_to_world(self, point: QPointF) -> tuple[float, float]:
        xmin, ymin, xmax, ymax = self._view_extent
        return (
            xmin + point.x() * (xmax - xmin) / max(1, self.width()),
            ymin + (self.height() - point.y()) * (ymax - ymin) / max(1, self.height()),
        )

    def _image_for(self, layer_id: str) -> QImage:
        assert self._scene is not None
        key = self._scene.scalar_raster_key(layer_id)
        cached = self._image_cache.get(layer_id)
        if cached is not None and cached[0] == key:
            return cached[1]
        rgba = self._scene.raster_rgba(layer_id)
        image = QImage(
            rgba.data,
            rgba.shape[1],
            rgba.shape[0],
            rgba.shape[1] * 4,
            QImage.Format.Format_RGBA8888,
        ).copy()
        self._image_cache[layer_id] = (key, image)
        return image

    @staticmethod
    def _color(values: tuple[int, int, int, int]) -> QColor:
        return QColor(*values)

    def paintEvent(self, _event) -> None:  # noqa: N802
        painter = QPainter(self)
        painter.fillRect(self.rect(), QColor(tokens.BG_SEARCH))
        if self._scene is None:
            painter.end()
            return
        for layer in self._scene.registry.layers():
            if not self._scene.registry.is_effectively_visible(layer.id, 1.0):
                continue
            scalar = self._scene.scalar_layer(layer.id)
            painter.save()
            painter.setOpacity(layer.opacity)
            if scalar is not None:
                xmin, ymin, xmax, ymax = layer.extent
                lower_left = self._world_to_screen(xmin, ymin)
                upper_right = self._world_to_screen(xmax, ymax)
                target = QRectF(upper_right, lower_left).normalized()
                painter.drawImage(target, self._image_for(layer.id))
            contour = self._scene.contour_geometry(layer.id)
            if contour is not None:
                painter.setPen(QPen(self._color(contour.color), contour.width))
                for path in contour.paths:
                    if len(path) < 2:
                        continue
                    for start, end in zip(path, path[1:]):
                        painter.drawLine(
                            self._world_to_screen(*start), self._world_to_screen(*end)
                        )
            points = self._scene.point_geometry(layer.id)
            if points is not None:
                painter.setPen(Qt.PenStyle.NoPen)
                painter.setBrush(self._color(points.color))
                for x, y in points.points:
                    center = self._world_to_screen(x, y)
                    painter.drawEllipse(center, points.radius, points.radius)
            painter.restore()
        painter.end()

    def wheelEvent(self, event: QWheelEvent) -> None:  # noqa: N802
        delta = event.angleDelta().y()
        if delta:
            self.zoom_by(0.8 if delta > 0 else 1.25, self._screen_to_world(event.position()))
            event.accept()
            return
        event.ignore()

    def mousePressEvent(self, event: QMouseEvent) -> None:  # noqa: N802
        if event.button() == Qt.MouseButton.LeftButton:
            self._last_drag_pos = event.position()
            self.setCursor(Qt.CursorShape.ClosedHandCursor)
            event.accept()
            return
        super().mousePressEvent(event)

    def mouseMoveEvent(self, event: QMouseEvent) -> None:  # noqa: N802
        if self._last_drag_pos is not None and event.buttons() & Qt.MouseButton.LeftButton:
            delta = event.position() - self._last_drag_pos
            self.pan_by_pixels(delta.x(), delta.y())
            self._last_drag_pos = event.position()
            event.accept()
            return
        super().mouseMoveEvent(event)

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:  # noqa: N802
        if event.button() == Qt.MouseButton.LeftButton and self._last_drag_pos is not None:
            self._last_drag_pos = None
            self.unsetCursor()
            event.accept()
            return
        super().mouseReleaseEvent(event)

File: libs/ui_canvas/src/qt/native_map_canvas.cpp
// UI-15 — native Qt map canvas (native_map_canvas.py parity).

#include <pwb/ui_canvas/qt/native_map_canvas.hpp>

#include <algorithm>
#include <cmath>
#include <set>
#include <stdexcept>

#include <QCloseEvent>
#include <QMouseEvent>
#include <QPainter>
#include <QPen>
#include <QWheelEvent>

#include <pwb/ui_widgets/ui_context.hpp>

namespace pwb::ui_canvas {

NativeMapCanvas::NativeMapCanvas(NativeMapScene* scene, QWidget* parent)
    : QWidget(parent) {
    setObjectName(QStringLiteral("NativeMapCanvas"));
    setMinimumSize(240, 180);
    setMouseTracking(true);
    connect(&raster_controller_,
            &NativeRasterRequestController::raster_ready, this,
            [this](const NativeRasterRequest& request,
                   const RasterImage& image) {
                on_raster_ready(request, image);
            });
    connect(&raster_controller_,
            &NativeRasterRequestController::raster_failed, this,
            [this](const NativeRasterRequest& request,
                   const QString& message) {
                on_raster_failed(request, message);
            });
    if (scene != nullptr) {
        set_scene(scene);
    }
}

NativeMapCanvas::~NativeMapCanvas() {
    if (scene_ != nullptr) {
        scene_->remove_change_listener(scene_listener_token_);
    }
}

void NativeMapCanvas::set_scene(NativeMapScene* scene) {
    if (scene_ != nullptr) {
        scene_->remove_change_listener(scene_listener_token_);
        scene_listener_token_ = 0;
    }
    scene_ = scene;
    ++scene_epoch_;
    raster_controller_.invalidate();
    if (scene_ != nullptr) {
        scene_listener_token_ = scene_->add_change_listener(
            [this]() { update(); });
    }
    image_cache_.clear();
    fit_to_scene();
    update();
}

void NativeMapCanvas::fit_to_scene() {
    if (scene_ == nullptr) {
        view_extent_ = {0.0, 0.0, 1.0, 1.0};
        return;
    }
    const std::array<double, 4> extent = scene_->extent();
    const double dx = extent[2] - extent[0];
    const double dy = extent[3] - extent[1];
    if (dx <= 0.0 || dy <= 0.0) {
        view_extent_ = {0.0, 0.0, 1.0, 1.0};
        return;
    }
    constexpr double margin = 0.04;
    view_extent_ = {extent[0] - dx * margin, extent[1] - dy * margin,
                    extent[2] + dx * margin, extent[3] + dy * margin};
}

void NativeMapCanvas::set_view_extent(
    const std::array<double, 4>& extent) {
    const double xmin = extent[0], ymin = extent[1];
    const double xmax = extent[2], ymax = extent[3];
    if (!(xmax > xmin && ymax > ymin)) {
        throw std::invalid_argument(
            "view extent must have positive width and height");
    }
    view_extent_ = {xmin, ymin, xmax, ymax};
    update();
}

void NativeMapCanvas::zoom_by(
    double factor, std::optional<std::pair<double, double>> center) {
    if (factor <= 0.0) {
        throw std::invalid_argument("zoom factor must be positive");
    }
    const double xmin = view_extent_[0], ymin = view_extent_[1];
    const double xmax = view_extent_[2], ymax = view_extent_[3];
    const auto [cx, cy] =
        center.value_or(
            std::make_pair((xmin + xmax) / 2.0, (ymin + ymax) / 2.0));
    set_view_extent({cx + (xmin - cx) * factor, cy + (ymin - cy) * factor,
                     cx + (xmax - cx) * factor, cy + (ymax - cy) * factor});
}

void NativeMapCanvas::pan_by_pixels(double dx, double dy) {
    const double xmin = view_extent_[0], ymin = view_extent_[1];
    const double xmax = view_extent_[2], ymax = view_extent_[3];
    const double width = std::max(1, this->width());
    const double height = std::max(1, this->height());
    const double world_dx = -dx * (xmax - xmin) / width;
    const double world_dy = dy * (ymax - ymin) / height;
    set_view_extent({xmin + world_dx, ymin + world_dy,
                     xmax + world_dx, ymax + world_dy});
}

QPointF NativeMapCanvas::world_to_screen(double x, double y) const {
    const double xmin = view_extent_[0], ymin = view_extent_[1];
    const double xmax = view_extent_[2], ymax = view_extent_[3];
    return QPointF((x - xmin) * width() / (xmax - xmin),
                   height() - (y - ymin) * height() / (ymax - ymin));
}

std::pair<double, double>
NativeMapCanvas::screen_to_world(const QPointF& point) const {
    const double xmin = view_extent_[0], ymin = view_extent_[1];
    const double xmax = view_extent_[2], ymax = view_extent_[3];
    const double w = std::max(1, width());
    const double h = std::max(1, height());
    return {xmin + point.x() * (xmax - xmin) / w,
            ymin + (height() - point.y()) * (ymax - ymin) / h};
}

QImage NativeMapCanvas::image_from_rgba(const RasterImage& raster) {
    // Copy — the QImage must own its pixels (RasterImage is transient).
    QImage image(raster.rgba.data(), raster.width, raster.height,
                 raster.stride, QImage::Format_RGBA8888);
    return image.copy();
}

QImage NativeMapCanvas::image_for(const std::string& layer_id) {
    Q_ASSERT(scene_ != nullptr);
    RasterKey key{0, 0};
    try {
        key = scene_->scalar_raster_key(layer_id);
    } catch (const std::out_of_range&) {
        return QImage();
    }
    const auto cached = image_cache_.find(layer_id);
    if (cached != image_cache_.end() && cached->second.first == key) {
        return cached->second.second;
    }
    ScalarRasterSourcePtr scalar = scene_->scalar_layer(layer_id);
    if (!scalar) {
        return QImage();
    }
    raster_controller_.request(scene_epoch_, layer_id, key,
                               std::move(scalar));
    return QImage();
}

void NativeMapCanvas::on_raster_ready(
    const NativeRasterRequest& request, const RasterImage& image) {
    if (scene_ == nullptr || request.scene_epoch != scene_epoch_) {
        return;
    }
    if (scene_->scalar_layer(request.layer_id) != request.scalar) {
        return;
    }
    try {
        if (scene_->scalar_raster_key(request.layer_id) !=
            request.raster_key) {
            return;
        }
    } catch (const std::out_of_range&) {
        return;
    }
    image_cache_[request.layer_id] =
        std::make_pair(request.raster_key, image_from_rgba(image));
    last_raster_error_.clear();
    update();
}

void NativeMapCanvas::on_raster_failed(
    const NativeRasterRequest& request, const QString& message) {
    const bool stale =
        scene_ != nullptr && request.scene_epoch != scene_epoch_;
    if (stale) {
        return;
    }
    if (scene_ != nullptr && request.scene_epoch == scene_epoch_) {
        last_raster_error_ = message.toStdString();
    }
    // The stored error previously had no consumer at all: a failed native
    // rasterization silently rendered as "no data" (#897). Log once per
    // layer+raster_key so repeated failures don't spam.
    const auto key =
        std::make_pair(request.layer_id, request.raster_key);
    if (logged_raster_errors_.insert(key).second) {
        qWarning("native scalar layer %s rasterization failed "
                 "(raster_key=(%llu,%llu)): %s",
                 request.layer_id.empty() ? "<unknown>"
                                          : request.layer_id.c_str(),
                 static_cast<unsigned long long>(request.raster_key.first),
                 static_cast<unsigned long long>(request.raster_key.second),
                 qUtf8Printable(message));
    }
}

void NativeMapCanvas::prepare_for_export() {
    if (scene_ == nullptr) {
        return;
    }
    for (const auto& layer : scene_->registry().layers()) {
        ScalarRasterSourcePtr scalar = scene_->scalar_layer(layer->id());
        if (!scalar) {
            continue;
        }
        RasterKey key{0, 0};
        try {
            key = scene_->scalar_raster_key(layer->id());
        } catch (const std::out_of_range&) {
            continue;
        }
        const auto cached = image_cache_.find(layer->id());
        if (cached != image_cache_.end() && cached->second.first == key) {
            continue;
        }
        image_cache_[layer->id()] = std::make_pair(
            key, image_from_rgba(scene_->raster_rgba(layer->id())));
    }
}

bool NativeMapCanvas::shutdown(int wait_ms) {
    return raster_controller_.shutdown(wait_ms);
}

const QImage& NativeMapCanvas::cached_image(
    const std::string& layer_id) const {
    static const QImage empty;
    const auto it = image_cache_.find(layer_id);
    return it == image_cache_.end() ? empty : it->second.second;
}

void NativeMapCanvas::paintEvent(QPaintEvent* /*event*/) {
    QPainter painter(this);
    painter.fillRect(rect(), QColor(pwb::ui_widgets::palette_token("BG_SEARCH")));
    if (scene_ == nullptr) {
        return;
    }
    for (const auto& layer : scene_->registry().layers()) {
        if (!scene_->registry().is_effectively_visible(layer->id(), 1.0)) {
            continue;
        }
        ScalarRasterSourcePtr scalar = scene_->scalar_layer(layer->id());
        painter.save();
        painter.setOpacity(layer->opacity());
        if (scalar) {
            const auto e = layer->extent();
            const QPointF lower_left = world_to_screen(e[0], e[1]);
            const QPointF upper_right = world_to_screen(e[2], e[3]);
            const QRectF target =
                QRectF(upper_right, lower_left).normalized();
            painter.drawImage(target, image_for(layer->id()));
        }
        if (const ContourGeometry* contour =
                scene_->contour_geometry(layer->id())) {
            painter.setPen(QPen(QColor(contour->color[0], contour->color[1],
                                       contour->color[2], contour->color[3]),
                                contour->width));
            for (const auto& path : contour->paths) {
                if (path.size() < 2) {
                    continue;
                }
                for (std::size_t i = 0; i + 1 < path.size(); ++i) {
                    painter.drawLine(world_to_screen(path[i].first,
                                                     path[i].second),
                                     world_to_screen(path[i + 1].first,
                                                     path[i + 1].second));
                }
            }
        }
        if (const PointGeometry* points =
                scene_->point_geometry(layer->id())) {
            painter.setPen(Qt::NoPen);
            painter.setBrush(QColor(points->color[0], points->color[1],
                                    points->color[2], points->color[3]));
            for (const auto& [x, y] : points->points) {
                painter.drawEllipse(world_to_screen(x, y), points->radius,
                                    points->radius);
            }
        }
        painter.restore();
    }
}

void NativeMapCanvas::wheelEvent(QWheelEvent* event) {
    const int delta = event->angleDelta().y();
    if (delta != 0) {
        const auto [wx, wy] = screen_to_world(event->position());
        zoom_by(delta > 0 ? 0.8 : 1.25, std::make_pair(wx, wy));
        event->accept();
        return;
    }
    event->ignore();
}

void NativeMapCanvas::mousePressEvent(QMouseEvent* event) {
    if (event->button() == Qt::LeftButton) {
        last_drag_pos_ = event->position();
        setCursor(Qt::ClosedHandCursor);
        event->accept();
        return;
    }
    QWidget::mousePressEvent(event);
}

void NativeMapCanvas::mouseMoveEvent(QMouseEvent* event) {
    if (last_drag_pos_.has_value() &&
        (event->buttons() & Qt::LeftButton)) {
        const QPointF delta = event->position() - *last_drag_pos_;
        pan_by_pixels(delta.x(), delta.y());
        last_drag_pos_ = event->position();
        event->accept();
        return;
    }
    QWidget::mouseMoveEvent(event);
}

void NativeMapCanvas::mouseReleaseEvent(QMouseEvent* event) {
    if (event->button() == Qt::LeftButton && last_drag_pos_.has_value()) {
        last_drag_pos_.reset();
        unsetCursor();
        event->accept();
        return;
    }
    QWidget::mouseReleaseEvent(event);
}

void NativeMapCanvas::closeEvent(QCloseEvent* event) {
    raster_controller_.shutdown();
    QWidget::closeEvent(event);
}

}  // namespace pwb::ui_canvas
